examples/java8/src/main/java/org/apache/beam/examples/snippets/Snippets.java

```diff
@@ -0,0 +1,385 @@
+
+      // [START BigQueryWriteFunction]
+      quotes.apply(BigQueryIO.<Quote>write()
+          .to(tableSpec)
+          .withSchema(tableSchema)
+          .withFormatFunction(
+              (Quote elem) -> new TableRow().set("source", elem.source).set("quote", elem.quote))
+          .withCreateDisposition(CreateDisposition.CREATE_IF_NEEDED)
+          .withWriteDisposition(WriteDisposition.WRITE_TRUNCATE));
+      // [END BigQueryWriteFunction]
+
+      // [START BigQueryWriteJsonSchema]
+      quotes.apply(BigQueryIO.<Quote>write()
+          .to(tableSpec)
+          .withJsonSchema(tableSchemaJson)
+          .withFormatFunction(
+              (Quote elem) -> new TableRow().set("source", elem.source).set("quote", elem.quote))
+          .withCreateDisposition(CreateDisposition.CREATE_IF_NEEDED)
+          .withWriteDisposition(WriteDisposition.WRITE_TRUNCATE));
+      // [END BigQueryWriteJsonSchema]
+    }
+
+    {
+      // [START BigQueryWriteDynamicDestinations]
+      /*
+      @DefaultCoder(AvroCoder.class)
+      static class WeatherData {
+        final long year;
+        final long month;
+        final long day;
+        final double maxTemp;
+
+        public WeatherData() {
+          this.year = 0;
+          this.month = 0;
+          this.day = 0;
+          this.maxTemp = 0.0f;
+        }
+        public WeatherData(long year, long month, long day, double maxTemp) {
+          this.year = year;
+          this.month = month;
+          this.day = day;
+          this.maxTemp = maxTemp;
+        }
+      }
+      */
+
+      PCollection<WeatherData> weatherData = p
+          .apply(BigQueryIO.read(
+              (SchemaAndRecord elem) -> {
+                GenericRecord record = elem.getRecord();
+                return new WeatherData(
+                    (Long) record.get("year"),
+                    (Long) record.get("month"),
+                    (Long) record.get("day"),
+                    (Double) record.get("max_temperature"));
+              })
+              .fromQuery("SELECT year, month, day, max_temperature "
+                  + "FROM [clouddataflow-readonly:samples.weather_stations] "
+                  + "WHERE year BETWEEN 2007 AND 2009")
+              .withCoder(AvroCoder.of(WeatherData.class)));
+
+      // We will send the weather data into different tables for every year.
+      weatherData.apply(BigQueryIO.<WeatherData>write()
+          .to(new DynamicDestinations<WeatherData, Long>() {
+            @Override
+            public Long getDestination(ValueInSingleWindow<WeatherData> elem) {
+              return elem.getValue().year;
+            }
+
+            @Override
+            public TableDestination getTable(Long destination) {
+              return new TableDestination(
+                  new TableReference()
+                      .setProjectId(writeProject)
+                      .setDatasetId(writeDataset)
+                      .setTableId(writeTable + "_" + destination),
```

Comment:
In terms of BigQuery best practices, partitioning by time is better done using partition decorators (`table$YYYYMMDD`).